images/virtualization-artifact/pkg/controller/vmip/internal/iplease_handler.go

```diff
@@ -0,0 +1,152 @@
+}
+
+func NewIPLeaseHandler(client client.Client, logger logr.Logger, ipAddressService *service.IpAddressService) *IPLeaseHandler {
+	return &IPLeaseHandler{
+		client:    client,
+		logger:    logger.WithValues("handler", IpLeaseHandlerName),
+		ipService: ipAddressService,
+	}
+}
+
+func (h IPLeaseHandler) Handle(ctx context.Context, state state.VMIPState) (reconcile.Result, error) {
+	vmip := state.VirtualMachineIP()
+	vmipStatus := state.VirtualMachineIPStatus()
+
+	lease, err := state.VirtualMachineIPLease(ctx)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	switch {
+	case lease == nil && vmipStatus.Address != "":
+		h.logger.Info("Lease by name not found: waiting for the lease to be available")
+		return reconcile.Result{}, nil
+
+	case lease == nil:
+		h.logger.Info("No Lease for VirtualMachineIP: create the new one", "type", vmip.Spec.Type, "address", vmip.Spec.StaticIP)
+		return h.createNewLease(ctx, state)
+
+	case lease.Status.Phase == "":
+		h.logger.Info("Lease is not ready: waiting for the lease")
+		return reconcile.Result{Requeue: true, RequeueAfter: 2 * time.Second}, nil
+
+	case util.IsBoundLease(lease, vmip):
+		h.logger.Info("Lease already exists, VirtualMachineIP ref is valid")
+		return reconcile.Result{}, nil
+
+	case lease.Status.Phase == virtv2.VirtualMachineIPAddressLeasePhaseBound:
+		h.logger.Info("Lease is bounded to another VirtualMachineIP: recreate VirtualMachineIP when the lease is released")
+		return reconcile.Result{}, nil
+
+	default:
+		h.logger.Info("Lease is released: set binding")
+
+		if lease.Spec.VirtualMachineIPAddressRef.Namespace != vmip.Namespace {
+			return reconcile.Result{}, fmt.Errorf("the selected VirtualMachineIP lease belongs to a different namespace: %s", lease.Spec.VirtualMachineIPAddressRef.Namespace)
+		}
+
+		lease.Spec.VirtualMachineIPAddressRef = &virtv2.VirtualMachineIPAddressLeaseIpAddressRef{
+			Name:      vmip.Name,
+			Namespace: vmip.Namespace,
+		}
+
+		err := h.client.Update(ctx, lease)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		vmipStatus.Address = common.LeaseNameToIP(lease.Name)
+		return reconcile.Result{}, nil
+	}
+}
+
+func (h IPLeaseHandler) createNewLease(ctx context.Context, state state.VMIPState) (reconcile.Result, error) {
+	vmip := state.VirtualMachineIP()
+	vmipStatus := state.VirtualMachineIPStatus()
+
+	if vmip.Spec.Type == virtv2.VirtualMachineIPAddressTypeAuto {
+		h.logger.Info("VirtualMachineIP address omitted in the spec: allocate the new one")
+		var err error
+		vmipStatus.Address, err = h.ipService.AllocateNewIP(state.AllocatedIPs())
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	} else {
+		vmipStatus.Address = vmip.Spec.StaticIP
+	}
+
+	if !h.ipService.IsAvailableAddress(vmipStatus.Address, state.AllocatedIPs()) {
+		h.logger.Info("VirtualMachineIP cannot be created: the address has already been allocated for another vmip", "address", vmip.Spec.StaticIP)
+		return reconcile.Result{}, nil
+	}
```

Comment:
This situation is lost in logs. Let's add Event for this situation. Also, check this case in lifecycle_handler, is it visible through conditions?

Reply:
Done